### PR TITLE
Keep out script

### PR DESCRIPTION
**Pull request by Misca1234**
Here is a script that create a "keep out area" based on arbitrarily closed polygon

	#
    # 1
	f.append(KeepOutArea(nodes=[[0, 0], [4, 0], [4, 4], [0, 4], [-2, 2]], width=0.4))
    #
    # 2
	f.append(KeepOutArea(nodes=[[10, 10], [14, 10], [14, 14], [10, 14]], width=0.01))
	#
    # 3
	f.append(KeepOutArea(nodes=[[20, 20], [24, 20], [24, 24], [22, 26], [20, 24]], layer="Cmts.User"))
    #
    # 4
	f.append(KeepOutArea(nodes=[[-25,-30], [5,-30], [5,-10], [0,-10], [0,-15], [-5,-15], [-5,-20], [-15,-20], [-15,-15], [-20,-15], [-20,-10], [-25,-10]]))
    #
    # 5
	f.append(KeepOutArea(nodes=[[15,-30], [20,-30], [20,-25], [25,-25], [25,-20], [30,-20], [35,-20], [35,-25], [40,-25], [40,-30], [45,-30], [45, -10], [15, -10]]))
    #
    # 6
	f.append(KeepOutArea(nodes=[[35,5], [37,3], [39,3], [45,10], [38,15], [37,12], [41,8], [36,5]]))

bild

### CODE REVIEW

File: KicadModTree/util/KeepOutArea.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        # regarding the distance from the intersection starting point
+        # Add them now as lines
+        #
+        c = 0
+        while ((c + 1) < len(cp)):
+            self._createChildNodes(cp[c][0], cp[c][1], cp[c+1][0], cp[c+1][1])
+            c = c + 2
+
+    def calculateBoundingBoxXX(self):
+        min_xy = Point2D(100000.0, 100000.0)
```
Comment (pointhi) [inline KicadModTree/util/KeepOutArea.py]: why those coordinates? Simply set to the first point at it will always work

**Comment by Misca1234** [comment]
An old habit, I can change it

**Comment by Misca1234** [comment]
Changed it to

`        min_xy = Point2D(0, 0)
        max_xy = Point2D(0, 0)

        if len(self.lines) > 0:
            min_xy = Point2D(self.lines[0].x, self.lines[0].y)
            max_xy = Point2D(self.lines[0].x, self.lines[0].y)
            `

**Comment by pointhi** [comment]
Please use https://github.com/pointhi/kicad-footprint-generator/blob/master/KicadModTree/PolygonPoints.py#L23 for the line as it is used in other nodes as well

**Comment by Misca1234** [comment]
The technique used is the straight line method, I need a class with k and m for each line internally when I calculate the intersections or are you referring to the final polygon ?

**Comment by Misca1234** [comment]
should this be moved to 
/tree/master/scripts/general
as well ?

**Comment by pointhi** [comment]
I would move it to ```KicadModTree/nodes/specialized/```. I thought about using PolygonePoint as input format, because it allows cutouts for example, and is used in other classes as well

**Comment by Misca1234** [comment]
I guess I use polygon points

 ```
 def __init__(self, **kwargs):
        Node.__init__(self)

        self.layer = kwargs.get('layer', 'F.SilkS')
        self.width = kwargs.get('width')

        self._initPolyPoint(**kwargs)

        self.virtual_childs = self._createChildNodes(self.nodes)

    def _initPolyPoint(self, **kwargs):
        self.nodes = PolygonPoints(**kwargs)
```


It have been moved as well

**Comment by Misca1234** [comment]
Could not recreate the branch locally, closing this one and re push it here
https://github.com/pointhi/kicad-footprint-generator/pull/242

**Pull request closed without merging**